File: backend/app/schemas/restaurant.py
"""Pydantic request/response models for the restaurants router."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

Tier = Literal["survive", "cost_effective", "luxury"]
VerificationStatus = Literal[
    "ai_parsed", "human_verified", "disputed", "needs_verification"
]
MenuStatus = Literal["populated_verified", "populated_ai", "empty"]


class CheapestMenu(BaseModel):
    id: UUID
    name: str
    price_cents: int
    tier: Tier
    verification_status: VerificationStatus


class NearbyRestaurant(BaseModel):
    id: UUID
    name: str
    category: str | None = None
    lat: float
    lng: float
    distance_m: int
    google_rating: float | None = None
    app_rating: float | None = None
    menu_status: MenuStatus
    cheapest_menu: CheapestMenu | None = None


class NearbyResponse(BaseModel):
    restaurants: list[NearbyRestaurant]
    count: int


class MenuItemOut(BaseModel):
    id: UUID
    name: str
    description: str | None = None
    price_cents: int
    photo_url: str | None = None
    verification_status: VerificationStatus
    confirmation_count: int
    source: str
    last_verified_at: datetime | None = None


class MenuByTier(BaseModel):
    survive: list[MenuItemOut] = Field(default_factory=list)
    cost_effective: list[MenuItemOut] = Field(default_factory=list)
    luxury: list[MenuItemOut] = Field(default_factory=list)


class RestaurantDetail(BaseModel):
    id: UUID
    name: str
    address: str | None = None
    phone: str | None = None
    website: str | None = None
    lat: float
    lng: float
    google_rating: float | None = None
    app_rating: float | None = None
    rating_count: int = 0
    hours: dict | None = None
    menu: MenuByTier

File: android/app/src/main/java/com/brokenlunch/gr/data/model/Restaurant.kt
package com.brokenlunch.gr.data.model

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class RestaurantNearby(
    @Json(name = "id") val id: String,
    @Json(name = "name") val name: String,
    @Json(name = "category") val category: String?,
    @Json(name = "lat") val lat: Double,
    @Json(name = "lng") val lng: Double,
    @Json(name = "distance_m") val distanceM: Int,
    @Json(name = "google_rating") val googleRating: Double?,
    @Json(name = "app_rating") val appRating: Double?,
    @Json(name = "menu_status") val menuStatus: MenuStatus,
    @Json(name = "cheapest_menu") val cheapestMenu: CheapestMenu?,
)

@JsonClass(generateAdapter = true)
data class CheapestMenu(
    @Json(name = "id") val id: String,
    @Json(name = "name") val name: String,
    @Json(name = "price_cents") val priceCents: Int,
    @Json(name = "tier") val tier: Tier,
    @Json(name = "verification_status") val verificationStatus: VerificationStatus,
)

@JsonClass(generateAdapter = true)
data class NearbyResponse(
    @Json(name = "restaurants") val restaurants: List<RestaurantNearby>,
    @Json(name = "count") val count: Int,
)

@JsonClass(generateAdapter = true)
data class RestaurantDetail(
    @Json(name = "id") val id: String,
    @Json(name = "name") val name: String,
    @Json(name = "address") val address: String?,
    @Json(name = "phone") val phone: String?,
    @Json(name = "website") val website: String?,
    @Json(name = "lat") val lat: Double,
    @Json(name = "lng") val lng: Double,
    @Json(name = "google_rating") val googleRating: Double?,
    @Json(name = "app_rating") val appRating: Double?,
    @Json(name = "rating_count") val ratingCount: Int,
    @Json(name = "menu") val menu: MenuByTier,
)

@JsonClass(generateAdapter = true)
data class MenuByTier(
    @Json(name = "survive") val survive: List<MenuItem> = emptyList(),
    @Json(name = "cost_effective") val costEffective: List<MenuItem> = emptyList(),
    @Json(name = "luxury") val luxury: List<MenuItem> = emptyList(),
) {
    fun isEmpty(): Boolean = survive.isEmpty() && costEffective.isEmpty() && luxury.isEmpty()
}
